src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -1284,7 +1423,65 @@ static void run_server(Preforker& forker, NBDServer *server, bool netlink_used)
   shutdown_async_signal_handler();
 }
 
-static int do_map(int argc, const char *argv[], Config *cfg)
+#ifndef __NR_pidfd_open
+#define __NR_pidfd_open 434   /* System call # on most architectures */
+#endif
+
+static int
+pidfd_open(pid_t pid, unsigned int flags)
+{
+  int fd = syscall(__NR_pidfd_open, pid, flags);
+
+  if (fd == -1 && errno == ENOSYS) {
+    // Fallback to open the /proc/pid directory
+
+    std::string path = "/proc/" + stringify(pid);
+    fd = open(path.c_str(), O_RDONLY);
+    if (fd == -1 && errno == ENOENT) {
+      errno = ESRCH;
+    }
+  }
+
+  return fd;
+}
+
+static int wait_for_terminate(int pid, int timeout)
+{
+  int fd = pidfd_open(pid, 0);
+  if (fd == -1) {
+    if (errno == -ESRCH) {
+      return 0;
+    }
+    int r = -errno;
+    cerr << "rbd-nbd: pidfd_open(" << pid << ") failed: "
+         << cpp_strerror(r) << std::endl;
+    return r;
+  }
+
+  struct pollfd poll_fds[1];
+  memset(poll_fds, 0, sizeof(struct pollfd));
+  poll_fds[0].fd = fd;
+  poll_fds[0].events = POLLIN;
+
+  int r = poll(poll_fds, 1, timeout * 1000);
+  if (r < 0) {
+    cerr << "rbd-nbd: failed to poll rbd-nbd process: " << cpp_strerror(r)
+         << std::endl;
+    goto done;
```

Comment:
Nit: leaking fd in this case

Reply:
Ah, actually the `done` label was introduced to avoid the leak, but I jut put it in a wrong place. Thanks.